exotations/solvers/exotica_control_rrt_solver/src/control_rrt_solver.cpp

```diff
@@ -0,0 +1,190 @@
+    setup_ = std::make_shared<oc::SimpleSetup>(cspace);
+    oc::SpaceInformationPtr si = setup_->getSpaceInformation();
+
+    si->setPropagationStepSize(dt);
+    setup_->setStateValidityChecker([this, si](const ob::State *state) { return isStateValid(si, state); });
+
+    std::shared_ptr<OMPLStatePropagator> propagator(
+        std::make_shared<OMPLStatePropagator>(si, dynamics_solver_));
+
+    setup_->setStatePropagator(propagator);
+
+    const Eigen::VectorXd start_eig = prob_->get_X().col(0);
+    const Eigen::MatrixXd goal_eig = prob_->get_X_star().col(T - 1);
+
+    start_state_ = std::make_shared<ob::ScopedState<ob::RealVectorStateSpace>>(space);
+    goal_state_ = std::make_shared<ob::ScopedState<ob::RealVectorStateSpace>>(space);
+
+    for (int i = 0; i < NX; ++i)
+    {
+        (*start_state_)[i] = start_eig(i);
+        (*goal_state_)[i] = goal_eig(i);
+    }
+
+    setup_->setStartAndGoalStates(*start_state_, *goal_state_, parameters_.ConvergenceTolerance);
+
+    ob::PlannerPtr optimizingPlanner(new ompl::control::RRT(si));
+    setup_->setPlanner(optimizingPlanner);
+
+    setup_->setup();
+    propagator->setIntegrationTimeStep(si->getPropagationStepSize());
+}
+
+void ControlRRTSolver::Solve(Eigen::MatrixXd &solution)
+{
+    if (!prob_) ThrowNamed("Solver has not been initialized!");
+    Timer planning_timer, backward_pass_timer, line_search_timer;
+
+    int T = prob_->get_T();
+    const int NU = prob_->get_num_controls();
+    const int NX = prob_->get_num_positions() + prob_->get_num_velocities();
+    const Eigen::VectorXd x_star = prob_->get_X_star().col(T - 1);
+
+    const double dt = dynamics_solver_->get_dt();
+
+    Setup();
+    ob::PlannerStatus solved = setup_->solve(parameters_.MaxIterationTime);
+
+    if (solved)
+    {
+        std::vector<oc::Control *> controls = setup_->getSolutionPath().getControls();
+        std::vector<ob::State *> states = setup_->getSolutionPath().getStates();
+        std::vector<double> durations = setup_->getSolutionPath().getControlDurations();
+
+        Eigen::VectorXd sol_goal_state = Eigen::Map<Eigen::VectorXd>(
+            states.back()->as<ob::RealVectorStateSpace::StateType>()->values, NX);
+
+        T = 0;
+        for (int t = 0; t < controls.size(); ++t)
+            T += (int)(durations[t] / dt);
+        T += 1;
+
+        if ((x_star - sol_goal_state).norm() > parameters_.ConvergenceTolerance && debug_)
+            WARNING_NAMED("ControlRRTSolver", "Goal not satisfied.");
+
+        // additional solution criteria
+        if (T <= 2 || ((x_star - sol_goal_state).norm() > parameters_.ConvergenceTolerance &&
+                       !parameters_.ApproximateSolution))
+        {
+            if (debug_) HIGHLIGHT_NAMED("ControlRRTSolver", "No solution found.");
+            return;
+        }
+        else if (debug_)
+            HIGHLIGHT_NAMED("ControlRRTSolver", "Found solution.");
+
+        prob_->set_T(T);
+        prob_->PreUpdate();
```

Comment:
I see why you are doing this, but this would not work if someone did change something about the problem (Q, Xstar, taskmaps which dont work here anyways). There is no way to do this without changing the problem inside the solver?

Reply:
I don't see how, since you don't know the number of timesteps before RRT converges.

Reply:
Yes, but instead of changing the problem you can use the start-state and integrator from dynamics-solver to do the roll-out just in the motion solver. No need to change problem.

Reply:
But then the solution will have a different `T` from the problem. Which will break attempts to playback the solution by calling `update` on the problem for all timesteps.

Reply:
True, but that has to be handled in user code I feel.

Reply:
I'm not convinced - wouldn't you want the problem to contain a solution once you've called `solve` on it?